File: demo/web-scienze-pronta/quiz-engine.js
(function (global) {
  class QuizEngine {
    constructor(questions) {
      this.allQuestions = Array.isArray(questions) ? questions : [];
      this.activeQuestions = [];
      this.currentIndex = 0;
      this.score = 0;
      this.answeredCount = 0;
    }

    startQuiz({ categoria = "tutte", livello = "tutti", numeroDomande = 10 } = {}) {
      const filtered = this.allQuestions
        .filter((question) => this.categoryMatches(question, categoria))
        .filter((question) => livello === "tutti" || question.livello === livello);

      this.activeQuestions = this.shuffle(filtered);

      if (numeroDomande > 0) {
        this.activeQuestions = this.activeQuestions.slice(0, numeroDomande);
      }

      this.currentIndex = 0;
      this.score = 0;
      this.answeredCount = 0;

      return this.activeQuestions;
    }

    currentQuestion() {
      return this.activeQuestions[this.currentIndex] || null;
    }

    answer(selectedAnswer) {
      const question = this.currentQuestion();

      if (!question) {
        throw new Error("Nessuna domanda attiva.");
      }

      const isCorrect = selectedAnswer === question.risposta_corretta;

      if (isCorrect) {
        this.score += 1;
      }

      this.answeredCount += 1;

      return {
        isCorrect,
        selectedAnswer,
        correctAnswer: question.risposta_corretta,
        explanation: question.spiegazione || "Spiegazione non disponibile.",
        score: this.score,
        totalAnswered: this.answeredCount,
      };
    }

    moveNext() {
      this.currentIndex += 1;
      return this.currentQuestion();
    }

    hasNext() {
      return this.currentIndex < this.activeQuestions.length - 1;
    }

    progressText() {
      if (!this.activeQuestions.length) {
        return "0/0";
      }

      return `${this.currentIndex + 1}/${this.activeQuestions.length}`;
    }

    categoryMatches(question, categoria) {
      const selected = this.slug(categoria);

      if (selected === "tutte") {
        return true;
      }

      const category = this.slug(question.categoria);
      const subcategory = this.slug(question.sottocategoria);
      const tags = Array.isArray(question.tags)
        ? question.tags.map((tag) => this.slug(tag))
        : [];

      if (selected === category || selected === subcategory || tags.includes(selected)) {
        return true;
      }

      if (selected === "fisica") {
        return subcategory.includes("fisica") || tags.some((tag) => tag.includes("fisica"));
      }

      if (selected === "chimica") {
        return subcategory.includes("chimica") || tags.some((tag) => tag.includes("chimica"));
      }

      if (selected === "biologia") {
        return subcategory.includes("biologia") || tags.some((tag) => tag.includes("biologia"));
      }

      return false;
    }

    shuffle(array) {
      const copy = [...array];

      for (let index = copy.length - 1; index > 0; index--) {
        const randomIndex = Math.floor(Math.random() * (index + 1));
        [copy[index], copy[randomIndex]] = [copy[randomIndex], copy[index]];
      }

      return copy;
    }

    slug(text) {
      return String(text || "")
        .trim()
        .toLowerCase()
        .replace(/[àá]/g, "a")
        .replace(/[èé]/g, "e")
        .replace(/[ìí]/g, "i")
        .replace(/[òó]/g, "o")
        .replace(/[ùú]/g, "u")
        .replace(/\s+/g, "_")
        .replace(/-/g, "_");
    }
  }

  class QuizQualityValidator {
    static validate(questions) {
      const issues = [];
      const ids = new Set();

      questions.forEach((question, index) => {
        const id = question.id || `QUESTION_${index + 1}`;

        if (!question.id) {
          issues.push({ questionId: id, severity: "warning", message: "ID mancante." });
        }

        if (question.id && ids.has(question.id)) {
          issues.push({ questionId: id, severity: "error", message: "ID duplicato." });
        }

        ids.add(question.id);

        if (!question.categoria) {
          issues.push({ questionId: id, severity: "error", message: "Categoria mancante." });
        }

        if (!question.livello) {
          issues.push({ questionId: id, severity: "error", message: "Livello mancante." });
        }

        if (!question.domanda) {
          issues.push({ questionId: id, severity: "error", message: "Testo domanda mancante." });
        }

        if (!Array.isArray(question.opzioni) || question.opzioni.length !== 4) {
          issues.push({ questionId: id, severity: "error", message: "La domanda deve avere 4 opzioni." });
        }

        if (Array.isArray(question.opzioni) && new Set(question.opzioni).size !== question.opzioni.length) {
          issues.push({ questionId: id, severity: "error", message: "Opzioni duplicate." });
        }

        if (!question.risposta_corretta) {
          issues.push({ questionId: id, severity: "error", message: "Risposta corretta mancante." });
        }

        if (Array.isArray(question.opzioni) && !question.opzioni.includes(question.risposta_corretta)) {
          issues.push({ questionId: id, severity: "error", message: "Risposta corretta non presente tra le opzioni." });
        }

        if (!question.spiegazione) {
          issues.push({ questionId: id, severity: "warning", message: "Spiegazione mancante." });
        }
      });

      return issues;
    }

    static hasBlockingErrors(questions) {
      return this.validate(questions).some((issue) => issue.severity === "error");
    }
  }

  global.AlexQuizEngine = {
    QuizEngine,
    QuizQualityValidator,
  };
})(window);

File: runtime/android/kotlin/com/alex/quizengine/QuizEngine.kt
package com.alex.quizengine

class QuizEngine(
    private val allQuestions: List<QuizQuestion>
) {
    private var activeQuestions: List<QuizQuestion> = emptyList()
    private var currentIndex: Int = 0
    private var score: Int = 0
    private var answeredCount: Int = 0
    private val selectedAnswers: MutableMap<String, String> = mutableMapOf()

    fun startQuiz(config: QuizConfig = QuizConfig()): List<QuizQuestion> {
        return startQuiz(
            categoria = config.categoria,
            livello = config.livello,
            numeroDomande = config.numeroDomande
        )
    }

    fun startQuiz(
        categoria: String = "tutte",
        livello: String = "tutti",
        numeroDomande: Int = 10
    ): List<QuizQuestion> {
        val filteredQuestions = allQuestions
            .filter { categoryMatches(it, categoria) }
            .filter { levelMatches(it, livello) }
            .shuffled()

        activeQuestions = if (numeroDomande <= 0) {
            filteredQuestions
        } else {
            filteredQuestions.take(numeroDomande)
        }

        currentIndex = 0
        score = 0
        answeredCount = 0
        selectedAnswers.clear()

        return activeQuestions
    }

    fun currentQuestion(): QuizQuestion? {
        return activeQuestions.getOrNull(currentIndex)
    }

    fun answer(selectedAnswer: String): AnswerResult {
        val question = currentQuestion()
            ?: throw IllegalStateException("Nessuna domanda attiva.")

        if (selectedAnswer !in question.opzioni) {
            throw IllegalArgumentException("La risposta selezionata non è tra le opzioni della domanda.")
        }

        val answerKey = question.id.ifBlank { "QUESTION_$currentIndex" }

        if (selectedAnswers.containsKey(answerKey)) {
            throw IllegalStateException("Questa domanda ha già ricevuto una risposta.")
        }

        val isCorrect = selectedAnswer == question.rispostaCorretta

        if (isCorrect) {
            score += 1
        }

        answeredCount += 1
        selectedAnswers[answerKey] = selectedAnswer

        return AnswerResult(
            isCorrect = isCorrect,
            selectedAnswer = selectedAnswer,
            correctAnswer = question.rispostaCorretta,
            explanation = question.spiegazione,
            score = score,
            totalAnswered = answeredCount,
            totalQuestions = activeQuestions.size
        )
    }

    fun moveNext(): QuizQuestion? {
        if (hasNext()) {
            currentIndex += 1
        }

        return currentQuestion()
    }

    fun hasNext(): Boolean {
        return currentIndex < activeQuestions.size - 1
    }

    fun isFinished(): Boolean {
        return activeQuestions.isNotEmpty() && answeredCount >= activeQuestions.size
    }

    fun totalQuestions(): Int {
        return activeQuestions.size
    }

    fun currentScore(): Int {
        return score
    }

    fun answeredQuestions(): Int {
        return answeredCount
    }

    fun progressText(): String {
        if (activeQuestions.isEmpty()) {
            return "0/0"
        }

        return "${currentIndex + 1}/${activeQuestions.size}"
    }

    fun summary(): QuizSummary {
        val total = activeQuestions.size
        val percentage = ScoreEngine.percentage(score, total)
        val label = ScoreEngine.label(score, total)

        return QuizSummary(
            score = score,
            totalQuestions = total,
            percentage = percentage,
            label = label,
            finalMessage = ScoreEngine.finalMessage(score, total)
        )
    }

    fun availableCategories(): List<String> {
        return allQuestions
            .map { it.categoria }
            .filter { it.isNotBlank() }
            .distinct()
            .sorted()
    }

    fun availableLevels(): List<String> {
        return allQuestions
            .map { it.livello }
            .filter { it.isNotBlank() }
            .distinct()
            .sorted()
    }

    private fun categoryMatches(question: QuizQuestion, categoria: String): Boolean {
        val selectedCategory = slug(categoria)

        if (selectedCategory == "tutte") {
            return true
        }

        val questionCategory = slug(question.categoria)
        val questionSubcategory = slug(question.sottocategoria)
        val questionTags = question.tags.map { slug(it) }

        if (
            selectedCategory == questionCategory ||
            selectedCategory == questionSubcategory ||
            selectedCategory in questionTags
        ) {
            return true
        }

        return when (selectedCategory) {
            "fisica" -> questionSubcategory.contains("fisica") ||
                questionTags.any { it.contains("fisica") }

            "chimica" -> questionSubcategory.contains("chimica") ||
                questionTags.any { it.contains("chimica") }

            "biologia" -> questionSubcategory.contains("biologia") ||
                questionTags.any { it.contains("biologia") }

            else -> false
        }
    }

    private fun levelMatches(question: QuizQuestion, livello: String): Boolean {
        val selectedLevel = slug(livello)

        if (selectedLevel == "tutti") {
            return true
        }

        return slug(question.livello) == selectedLevel
    }

    private fun slug(text: String): String {
        return text
            .trim()
            .lowercase()
            .replace("à", "a")
            .replace("è", "e")
            .replace("é", "e")
            .replace("ì", "i")
            .replace("ò", "o")
            .replace("ù", "u")
            .replace(" ", "_")
            .replace("-", "_")
    }
}
